lxd/storage_volumes_snapshot.go

```diff
@@ -724,13 +700,72 @@ func autoCreateCustomVolumeSnapshotsTask(d *Daemon) (task.Func, task.Schedule) {
 				continue
 			}
 
-			volumes = append(volumes, v)
+			// Always include local volumes
+			if v.NodeID == nodeID {
+				volumes = append(volumes, v)
+			} else if v.NodeID < 0 {
+				// Keep a separat list of remote volumes
+				remoteVolumes = append(remoteVolumes, v)
+			}
 		}
 
-		if len(volumes) == 0 {
+		// Exit early if there are no volumes to be snapshotted.
+		if len(volumes) == 0 && len(remoteVolumes) == 0 {
 			return
 		}
 
+		var nodes []api.ClusterMember
+		var availableNodeIDs []int64
+
+		if len(remoteVolumes) > 0 {
+			// Get list of cluster nodes
+			nodes, err = cluster.List(d.State(), d.gateway)
+			if err != nil {
+				return
+			}
+
+			for _, node := range nodes {
+				var nodeID int64
+
+				if node.Status == "Online" {
+					err := d.cluster.Transaction(func(tx *db.ClusterTx) error {
+						info, err := tx.GetNodeByName(node.ServerName)
+						if err != nil {
+							return err
+						}
+
+						nodeID = info.ID
+						return nil
+					})
+					if err != nil {
+						return
+					}
+
+					availableNodeIDs = append(availableNodeIDs, nodeID)
+				}
+			}
+
+			if len(availableNodeIDs) > 1 {
+				// Select a stable random node to perform the snapshot.
+				for _, v := range remoteVolumes {
+					selectedNodeID, err := util.GetStableRandomInt64FromList(int64(v.ID), availableNodeIDs)
+					if err != nil {
+						continue
+					}
+
+					// Don't snapshot, if we're not the chosen one.
+					if nodeID != selectedNodeID {
+						continue
+					}
+
+					volumes = append(volumes, v)
+				}
+			} else {
+				// Let the local node perform the snapshot as there are no other cluster members.
+				volumes = append(volumes, remoteVolumes...)
```

Comment:
@monstermunchkin I wonder whether it is safe/proper to perform the snapshots on potentially all nodes, if they are all isolated from one and other (such that no node thinks any of the others are online).

Maybe we should differentiate between a single non-clustered node that uses remote volumes, vs a clustered node that uses remote volumes when the cluster is partitioned? We could use `clustered, err := cluster.Enabled(d.db)` for the differentiation. CC @stgraber

Reply:
We shouldn't have all nodes perform snapshots for remote volumes. That was an issue in the past IIRC.

> Maybe we should differentiate between a single non-clustered node that uses remote volumes, vs a clustered node that uses remote volumes when the cluster is partitioned?

I don't understand why we'd need to do that. The code ensures that only one node performs a remote volume snapshot which is what we want.

Reply:
@monstermunchkin indeed, but my concern here was, is it possible for `nodes, err = cluster.List(d.State(), d.gateway)` to return a nil-error and a list of nodes that contains none with an "Online" status (perhaps in the scenario where the local node is isolated from the rest of the cluster)?

In that case, the remote snapshot would then end up being attempted on both the isolated node and one of the nodes in the rest of the "online" cluster.

I'm not sure if this is possible, but I thought it might be worth differentiating between a standalone node using `cluster.Enabled()` before calling `cluster.List()` and in the case where the node is clustered, but there are 0 online nodes, then skip taking snapshots, as something is clearly very wrong with the cluster.

Reply:
@tomponline Is the recent change what you had in mind?

Reply:
@monstermunchkin yeah this is good thanks. 

Where you have below, it would be good to log an error saying "Skipping auto custom volume snapshots: %v" or similar so its logged.

```
if err != nil {
    return
}
```

As for the `len(availableNodeIDs) < 2 {` bit, I guess we don't know for sure if the list contains a single online node that it is infact our own node, I think I would be tempted to just skip snapshotting and log an error if `len(availableNodeIDs) <= 0`, and only proceed to pick a snapshot node if len(availableNodeIDs) > 0 and is clustered.